lmcache/cache_engine.py

```diff
@@ -380,6 +406,49 @@ def retrieve(
 
         return ret, ret_mask
 
+    @_lmcache_nvtx_annotate
+    @torch.no_grad()
+    def lookup(
+        self,
+        tokens: torch.Tensor,
+        align_to_chunk: bool,
+    ) -> torch.Tensor:
+        """
+        Checks the existence of KV cache of the tokens from the cache engine.
+        The return should be a boolean tensor indicating the existence of 
+        the KV cache of each token.
+
+        Input:
+            tokens: the input tokens, with shape [seq_len]
+            align_to_chunk: whether to align the output to the chunk size, 
+            this truncates available tokens to the last full chunk.
+
+        Output:
+            a boolean tensor indicating the existence of the KV cache of 
+            each token
+        """
+        # NOTE(Sixian): Now this is a prefix lookup.
+        fmt = self.metadata.fmt
+        total_token_cnt = len(tokens)
+        ret = torch.zeros(total_token_cnt, dtype=torch.bool)
+        current_token_idx = 0
+        full_chunk_cnt = total_token_cnt // self.chunk_size
+        next_token_idx = total_token_cnt if full_chunk_cnt == 0 \
+            else self.chunk_size
+        chunk_hashes = self._prefix_hash(self._chunk_tokens(tokens), 0)
+        for chunk_hash in chunk_hashes:
+            interval_len = next_token_idx - current_token_idx
+            if interval_len < self.chunk_size and align_to_chunk:
+                break
+            if not self.engine_.contains(self._make_key(chunk_hash, fmt)):
+                break
+            else:
+                ret[current_token_idx:next_token_idx] = True
+            current_token_idx = next_token_idx
+            next_token_idx = min(next_token_idx + self.chunk_size,
+                                 total_token_cnt)
```

Comment:
The logic in this for-loop seems a bit complicated. Can you explain it? Maybe we can try making it easier to understand